Submitted By: Robert Connolly <robert at linuxfromscratch dot org> (ashes)
Date: 2004-04-18
Initial Package Version: binutils-2.15.90.0.3
Origin: http://pax.grsecurity.net/\
	binutils-2.14.90.0.8-pt-pax-flags-200402042140.patch
Description: This is needed for Pax. http://pax.grsecurity.net/
http://www.linuxfromscratch.org/hlfs/

diff -Naur binutils-2.15.90.0.3.orig/bfd/elf-bfd.h binutils-2.15.90.0.3.pt_pax/bfd/elf-bfd.h
--- binutils-2.15.90.0.3.orig/bfd/elf-bfd.h	2004-04-12 19:56:33.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/bfd/elf-bfd.h	2004-04-18 04:48:51.000000000 +0000
@@ -1204,6 +1204,9 @@
   /* Segment flags for the PT_GNU_STACK segment.  */
   unsigned int stack_flags;  
 
+  /* Segment flags for the PT_PAX_FLAGS segment.  */
+  unsigned int pax_flags;
+
   /* Symbol version definitions in external objects.  */
   Elf_Internal_Verdef *verdef;
 
diff -Naur binutils-2.15.90.0.3.orig/bfd/elf.c binutils-2.15.90.0.3.pt_pax/bfd/elf.c
--- binutils-2.15.90.0.3.orig/bfd/elf.c	2004-04-12 19:56:33.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/bfd/elf.c	2004-04-18 04:48:51.000000000 +0000
@@ -968,6 +968,7 @@
 	    case PT_TLS: pt = "TLS"; break;
 	    case PT_GNU_EH_FRAME: pt = "EH_FRAME"; break;
 	    case PT_GNU_STACK: pt = "STACK"; break;
+	    case PT_PAX_FLAGS: pt = "PAX_FLAGS"; break;
 	    default: sprintf (buf, "0x%lx", p->p_type); pt = buf; break;
 	    }
 	  fprintf (f, "%8s off    0x", pt);
@@ -2291,6 +2292,9 @@
     case PT_GNU_STACK:
       return _bfd_elf_make_section_from_phdr (abfd, hdr, index, "stack");
 
+    case PT_PAX_FLAGS:
+      return _bfd_elf_make_section_from_phdr (abfd, hdr, index, "pax_flags");
+
     default:
       /* Check for any processor-specific program segment types.
          If no handler for them, default to making "segment" sections.  */
@@ -3514,6 +3518,20 @@
       pm = &m->next;
     }
 
+    {
+      amt = sizeof (struct elf_segment_map);
+      m = bfd_zalloc (abfd, amt);
+      if (m == NULL)
+	goto error_return;
+      m->next = NULL;
+      m->p_type = PT_PAX_FLAGS;
+      m->p_flags = elf_tdata (abfd)->pax_flags;
+      m->p_flags_valid = 1;
+
+      *pm = m;
+      pm = &m->next;
+    }
+
   free (sections);
   sections = NULL;
 
@@ -4135,6 +4153,11 @@
       ++segs;
     }
 
+    {
+      /* We need a PT_PAX_FLAGS segment.  */
+      ++segs;
+    }
+
   for (s = abfd->sections; s != NULL; s = s->next)
     {
       if ((s->flags & SEC_LOAD) != 0
@@ -4684,6 +4707,7 @@
     || IS_COREFILE_NOTE (segment, section))				\
    && section->output_section != NULL					\
    && segment->p_type != PT_GNU_STACK					\
+   && segment->p_type != PT_PAX_FLAGS					\
    && (segment->p_type != PT_TLS					\
        || (section->flags & SEC_THREAD_LOCAL))				\
    && (segment->p_type == PT_LOAD					\
diff -Naur binutils-2.15.90.0.3.orig/bfd/elflink.c binutils-2.15.90.0.3.pt_pax/bfd/elflink.c
--- binutils-2.15.90.0.3.orig/bfd/elflink.c	2004-04-15 04:26:05.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/bfd/elflink.c	2004-04-18 05:45:55.000000000 +0000
@@ -4621,16 +4621,30 @@
   if (!is_elf_hash_table (info->hash))
     return TRUE;
 
+  elf_tdata (output_bfd)->pax_flags = PF_NORANDEXEC;
+
+  if (info->execheap)
+    elf_tdata (output_bfd)->pax_flags |= PF_NOMPROTECT;
+  else if (info->noexecheap)
+    elf_tdata (output_bfd)->pax_flags |= PF_MPROTECT;
+
   if (info->execstack)
-    elf_tdata (output_bfd)->stack_flags = PF_R | PF_W | PF_X;
+    {
+      elf_tdata (output_bfd)->stack_flags = PF_R | PF_W | PF_X;
+      elf_tdata (output_bfd)->pax_flags |= PF_EMUTRAMP;
+    }
   else if (info->noexecstack)
-    elf_tdata (output_bfd)->stack_flags = PF_R | PF_W;
+    {
+      elf_tdata (output_bfd)->stack_flags = PF_R | PF_W;
+      elf_tdata (output_bfd)->pax_flags |= PF_NOEMUTRAMP;
+    }
   else
     {
       bfd *inputobj;
       asection *notesec = NULL;
       int exec = 0;
 
+      elf_tdata (output_bfd)->pax_flags |= PF_NOEMUTRAMP;
       for (inputobj = info->input_bfds;
 	   inputobj;
 	   inputobj = inputobj->link_next)
@@ -4643,7 +4657,11 @@
 	  if (s)
 	    {
 	      if (s->flags & SEC_CODE)
-		exec = PF_X;
+               {
+                 elf_tdata (output_bfd)->pax_flags &= ~PF_NOEMUTRAMP;
+                 elf_tdata (output_bfd)->pax_flags |= PF_EMUTRAMP;
+                 exec = PF_X;
+               }
 	      notesec = s;
 	    }
 	  else
diff -Naur binutils-2.15.90.0.3.orig/binutils/readelf.c binutils-2.15.90.0.3.pt_pax/binutils/readelf.c
--- binutils-2.15.90.0.3.orig/binutils/readelf.c	2004-04-12 19:56:34.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/binutils/readelf.c	2004-04-18 04:48:55.000000000 +0000
@@ -2171,6 +2171,7 @@
     case PT_GNU_EH_FRAME:
 			return "GNU_EH_FRAME";
     case PT_GNU_STACK:	return "STACK";
+    case PT_PAX_FLAGS:	return "PAX_FLAGS";
 
     default:
       if ((p_type >= PT_LOPROC) && (p_type <= PT_HIPROC))
diff -Naur binutils-2.15.90.0.3.orig/include/bfdlink.h binutils-2.15.90.0.3.pt_pax/include/bfdlink.h
--- binutils-2.15.90.0.3.orig/include/bfdlink.h	2004-04-12 19:56:35.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/include/bfdlink.h	2004-04-18 04:48:55.000000000 +0000
@@ -302,6 +302,14 @@
      flags.  */
   unsigned int noexecstack: 1;
 
+  /* TRUE if PT_PAX_FLAGS segment should be created with PF_NOMPROTECT
+     flags.  */
+  unsigned int execheap: 1;
+
+  /* TRUE if PT_PAX_FLAGS segment should be created with PF_MPROTECT
+     flags.  */
+  unsigned int noexecheap: 1;
+
   /* What to do with unresolved symbols in an object file.
      When producing executables the default is GENERATE_ERROR.
      When producing shared libraries the default is IGNORE.  The
diff -Naur binutils-2.15.90.0.3.orig/include/elf/common.h binutils-2.15.90.0.3.pt_pax/include/elf/common.h
--- binutils-2.15.90.0.3.orig/include/elf/common.h	2004-04-12 19:56:35.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/include/elf/common.h	2004-04-18 04:48:55.000000000 +0000
@@ -290,12 +290,27 @@
 
 #define PT_GNU_EH_FRAME	(PT_LOOS + 0x474e550)
 #define PT_GNU_STACK	(PT_LOOS + 0x474e551)
+#define PT_PAX_FLAGS	(PT_LOOS + 0x5041580)
 
 /* Program segment permissions, in program header p_flags field.  */
 
 #define PF_X		(1 << 0)	/* Segment is executable */
 #define PF_W		(1 << 1)	/* Segment is writable */
 #define PF_R		(1 << 2)	/* Segment is readable */
+
+#define PF_PAGEEXEC	(1 << 4)	/* Enable  PAGEEXEC */
+#define PF_NOPAGEEXEC	(1 << 5)	/* Disable PAGEEXEC */
+#define PF_SEGMEXEC	(1 << 6)	/* Enable  SEGMEXEC */
+#define PF_NOSEGMEXEC	(1 << 7)	/* Disable SEGMEXEC */
+#define PF_MPROTECT	(1 << 8)	/* Enable  MPROTECT */
+#define PF_NOMPROTECT	(1 << 9)	/* Disable MPROTECT */
+#define PF_RANDEXEC	(1 << 10)	/* Enable  RANDEXEC */
+#define PF_NORANDEXEC	(1 << 11)	/* Disable RANDEXEC */
+#define PF_EMUTRAMP	(1 << 12)	/* Enable  EMUTRAMP */
+#define PF_NOEMUTRAMP	(1 << 13)	/* Disable EMUTRAMP */
+#define PF_RANDMMAP	(1 << 14)	/* Enable  RANDMMAP */
+#define PF_NORANDMMAP	(1 << 15)	/* Disable RANDMMAP */
+
 /* #define PF_MASKOS	0x0F000000    *//* OS-specific reserved bits */
 #define PF_MASKOS	0x0FF00000	/* New value, Oct 4, 1999 Draft */
 #define PF_MASKPROC	0xF0000000	/* Processor-specific reserved bits */
diff -Naur binutils-2.15.90.0.3.orig/ld/emultempl/elf32.em binutils-2.15.90.0.3.pt_pax/ld/emultempl/elf32.em
--- binutils-2.15.90.0.3.orig/ld/emultempl/elf32.em	2004-04-12 19:56:36.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/ld/emultempl/elf32.em	2004-04-18 04:48:55.000000000 +0000
@@ -1666,6 +1666,16 @@
 	  link_info.noexecstack = TRUE;
 	  link_info.execstack = FALSE;
 	}
+      else if (strcmp (optarg, "execheap") == 0)
+	{
+	  link_info.execheap = TRUE;
+	  link_info.noexecheap = FALSE;
+	}
+      else if (strcmp (optarg, "noexecheap") == 0)
+	{
+	  link_info.noexecheap = TRUE;
+	  link_info.execheap = FALSE;
+	}
       /* What about the other Solaris -z options? FIXME.  */
       break;
 EOF
@@ -1702,6 +1712,7 @@
   fprintf (file, _("  -z combreloc\t\tMerge dynamic relocs into one section and sort\n"));
   fprintf (file, _("  -z defs\t\tReport unresolved symbols in object files.\n"));
   fprintf (file, _("  -z execstack\t\tMark executable as requiring executable stack\n"));
+  fprintf (file, _("  -z execheap\t\tMark executable as requiring executable heap\n"));
   fprintf (file, _("  -z initfirst\t\tMark DSO to be initialized first at runtime\n"));
   fprintf (file, _("  -z interpose\t\tMark object to interpose all DSOs but executable\n"));
   fprintf (file, _("  -z loadfltr\t\tMark object requiring immediate process\n"));
@@ -1713,6 +1724,7 @@
   fprintf (file, _("  -z nodlopen\t\tMark DSO not available to dlopen\n"));
   fprintf (file, _("  -z nodump\t\tMark DSO not available to dldump\n"));
   fprintf (file, _("  -z noexecstack\tMark executable as not requiring executable stack\n"));
+  fprintf (file, _("  -z noexecheap\t\tMark executable as not requiring executable heap\n"));
   fprintf (file, _("  -z now\t\tMark object non-lazy runtime binding\n"));
   fprintf (file, _("  -z origin\t\tMark object requiring immediate \$ORIGIN processing\n\t\t\t  at runtime\n"));
   fprintf (file, _("  -z KEYWORD\t\tIgnored for Solaris compatibility\n"));
diff -Naur binutils-2.15.90.0.3.orig/ld/ldgram.y binutils-2.15.90.0.3.pt_pax/ld/ldgram.y
--- binutils-2.15.90.0.3.orig/ld/ldgram.y	2004-03-03 20:24:34.000000000 +0000
+++ binutils-2.15.90.0.3.pt_pax/ld/ldgram.y	2004-04-18 04:48:55.000000000 +0000
@@ -1019,6 +1019,8 @@
 			    $$ = exp_intop (0x6474e550);
 			  else if (strcmp (s, "PT_GNU_STACK") == 0)
 			    $$ = exp_intop (0x6474e551);
+			  else if (strcmp (s, "PT_PAX_FLAGS") == 0)
+			    $$ = exp_intop (0x65041580);
 			  else
 			    {
 			      einfo (_("\
